ibis/backends/dask/udf.py

```diff
@@ -0,0 +1,86 @@
+import itertools
+
+import dask.dataframe as dd
+import dask.dataframe.groupby as ddgb
+import pandas
+
+import ibis.client
+import ibis.expr.operations as ops
+import ibis.udf.vectorized
+from ibis.backends.pandas.udf import nullable, udf  # noqa
+
+from .dispatch import execute_node, pre_execute
+from .execution.util import make_selected_obj
+
+
+@pre_execute.register(ops.ElementWiseVectorizedUDF)
+@pre_execute.register(ops.ElementWiseVectorizedUDF, ibis.client.Client)
+def pre_execute_elementwise_udf(op, *clients, scope=None, **kwargs):
+    """Register execution rules for elementwise UDFs.
+    """
+    input_type = op.input_type
+
+    # definitions
+
+    # Define an execution rule for elementwise operations on a
+    # grouped Series
+    nargs = len(input_type)
+
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF,
+        *(itertools.repeat(ddgb.SeriesGroupBy, nargs)),
+    )
+    def execute_udf_node_groupby(op, *args, **kwargs):
+        func = op.func
+
+        # all grouping keys must be identical
+        assert len(set(tuple(arg.index) for arg in args)) == 1
+
+        # we're performing a scalar operation on grouped column, so
+        # perform the operation directly on the underlying Series
+        # and regroup after it's finished
+        args_objs = [make_selected_obj(arg) for arg in args]
+        groupings = args[0].index
+        return func(*args_objs).groupby(groupings)
+
+    # Define an execution rule for a simple elementwise Series
+    # function
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(pandas.Series, nargs))
+    )
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(dd.Series, nargs))
+    )
+    @execute_node.register(
+        ops.ElementWiseVectorizedUDF, *(itertools.repeat(object, nargs))
+    )
+    def execute_udf_node(op, *args, **kwargs):
+        # We have rewritten op.func to be a closure enclosing
+        # the kwargs, and therefore, we do not need to pass
+        # kwargs here. This is true for all udf execution in this
+        # file.
+        # See ibis.udf.vectorized.UserDefinedFunction
+        return op.func(*args)
+
+    return scope
+
+
+@pre_execute.register(ops.AnalyticVectorizedUDF)
+@pre_execute.register(ops.AnalyticVectorizedUDF, ibis.client.Client)
+@pre_execute.register(ops.ReductionVectorizedUDF)
+@pre_execute.register(ops.ReductionVectorizedUDF, ibis.client.Client)
+def pre_execute_analytic_and_reduction_udf(op, *clients, scope=None, **kwargs):
+    input_type = op.input_type
+    nargs = len(input_type)
+
+    # An execution rule to handle analytic and reduction UDFs over
+    # 1) an ungrouped window,
+    # 2) an ungrouped Aggregate node, or
+    # 3) an ungrouped custom aggregation context
+    @execute_node.register(type(op), *(itertools.repeat(dd.Series, nargs)))
```

Comment:
Why do we need both pandas.Series and dd.Series here?

Reply:
This was more out of an abundance of caution - now that we no longer share dispatchers with the pandas backend I wanted to make sure the series registration was also being made. I don't believe there's any real scenario where we're mixing dask Series and pandas Series objects through - so I could remove.

Reply:
yeah prefer to remove for clarify.